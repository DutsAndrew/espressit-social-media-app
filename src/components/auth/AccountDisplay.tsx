import React, { FC, useEffect, useState } from "react";
import { AccountDisplayProps, UserInstance } from "../../types/interfaces";
import chevron from '../../assets/chevron-down.svg';
import '../../styles/account.css';
import { User } from "firebase/auth";
import { doc, getDoc } from "firebase/firestore";
import { initializeApp } from "firebase/app";
import { getFirestore } from "firebase/firestore";
import anonymousProfile from '../../assets/profile-anonymous.svg';

const AccountDisplay: FC<AccountDisplayProps> = (props): JSX.Element => {

  const { 
    currentUser,
    signOut,
    toggleEditProfilePage,
    toggleViewFavoritesPage 
  } = props;

  // converting potential non-auth user to guaranteed firebase auth user
  const userRef = currentUser as User;

  const [userInstance, setUserInstance] = useState<UserInstance>({
    user: {
      comments: [],
      displayName: "",
      favoritePosts: [],
      posts: [],
      imgURL: "",
      uid: "",
      username: "",
    },
  });

  useEffect(() => {
    (async function fetchUserInstance() {

      //firebase config
      const firebaseConfig = {
        apiKey: "AIzaSyDsPecBa3Ch5uDw4UzHiJWAjKEYOKCrNdA",
        authDomain: "espressit.firebaseapp.com",
        projectId: "espressit",
        storageBucket: "espressit.appspot.com",
        messagingSenderId: "1094129721341",
        appId: "1:1094129721341:web:dc2bdc0a2b322504b04394"
      };
      // Initialize Firebase
      const app = initializeApp(firebaseConfig);
      const db = getFirestore(app);
  
      // fetch userInstance
      const userInstanceRef = doc(db, "users", userRef.uid);
      const userInstanceSnap = await getDoc(userInstanceRef);
      if (userInstanceSnap.exists()) {
        const userInstanceData = userInstanceSnap.data();
  
        // sync local state with user data
        setUserInstance({
          user: {
            comments: userInstanceData.comments,
            displayName: userInstanceData.displayName,
            favoritePosts: userInstanceData.favoritePosts,
            posts: userInstanceData.posts,
            imgURL: userInstanceData.imgURL,
            uid: userInstanceData.uid,
            username: userInstanceData.username,
          },
        });
      };
    })();
  });

  const accountDropDown = (e: React.MouseEvent) => {

    // open and closes drop down menu
    const profileContainer = document.querySelector('.profile-container');
    const dropDownMenu = document.querySelector('.account-drop-down-menu');
    const chevronButton = document.querySelector('.account-menu-button');
    const profileText = document.querySelector('.profile-text');

    if (!dropDownMenu) {
      chevronButton?.classList.add('drop-down-active');
      profileText?.classList.add('drop-down-text-active');
      const dropDownMenu = document.createElement('div');
        dropDownMenu.classList.remove('account-menu-button');
        dropDownMenu.classList.add('account-drop-down-menu');
      const signOut = document.createElement('p');
        signOut.classList.add('sign-out-text');
        signOut.textContent = 'Sign Out';
      const editProfile = document.createElement('p');
        editProfile.classList.add('edit-profile-text');
        editProfile.textContent = 'Edit Profile';
      const favoritePosts = document.createElement('p');
        favoritePosts.classList.add('view-favorite-text');
        favoritePosts.textContent = 'View Favorites';

      dropDownMenu.appendChild(signOut);
      dropDownMenu.appendChild(editProfile);
      dropDownMenu.appendChild(favoritePosts);
      profileContainer?.appendChild(dropDownMenu);

      setTimeout(() => {
        dropDownMenu?.remove()
        chevronButton?.classList.remove('drop-down-active');
        chevronButton?.classList.add('account-menu-button');
        profileText?.classList.remove('drop-down-text-active');
      }, 5000);
    };

    if (dropDownMenu) {
      dropDownMenu?.remove()
      chevronButton?.classList.remove('drop-down-active');
      chevronButton?.classList.add('account-menu-button');
      profileText?.classList.remove('drop-down-text-active');
    };


    const target = e.target as Element;

    if (target.classList.contains('sign-out-text')) {
      signOut();
      return;
    };

    if (target.classList.contains('edit-profile-text')) {
      toggleEditProfilePage();
      return;
    };

    if (target.classList.contains('view-favorite-text')) {
      toggleViewFavoritesPage();
      return;
    }

  };

  if (userInstance.user) {
    return (
      <div className="profile-container"
        onClick={accountDropDown}
        role="menu"
      >
        <p className="logged-in-as-text"
          role="paragraph">
          Logged in as:
        </p>
        <div className="img-name-container">
          <img className="profile-img"
            referrerPolicy="no-referrer"
            src={userInstance.user.imgURL ? userInstance.user.imgURL : anonymousProfile}
            alt="profile" >
          </img>
          <p className="profile-text" >
            {userInstance.user.username}
          </p>
          <img className="account-menu-button"
            src={chevron}
            alt="chevron" >
          </img>
        </div>
      </div>
    );
  } else {
    return (
      <div className="profile-container" 
        onClick={accountDropDown}
        role="menu"
      >
        <p className="logged-in-as-text"
          role="paragraph" >
          Logged in as:
        </p>
        <div className="img-name-container">
          <p className="profile-text">
            {userRef.email}
          </p>
          <img className="account-menu-button"
            src={chevron}
            alt="chevron" >
          </img>
        </div>
      </div>
    );
  };

};

export default AccountDisplay;