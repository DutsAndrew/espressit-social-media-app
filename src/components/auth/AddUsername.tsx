import { create } from "domain";
import React, { FC } from "react";
import { AddUsernameProps } from "types/interfaces";
import '../../styles/auth/AddUsername.css';
const Filter = require('bad-words');

const AddUsername: FC<AddUsernameProps> = (props): JSX.Element => {

  const { createUserInstanceAfterGoogleSignIn, toggleAddUsernamePage } = props;

  const filter = new Filter(),
        usernameFormat: RegExp = /^[a-z]{3,12}$|^[a-z]{3,12}\d{2,4}$/g;

  const validateUsername = (): void => {
    let input = (document.querySelector('.input-add-username') as HTMLInputElement).value;
    if (filter.isProfane(input)) {
      alert('we don\'t allow profane language');
      input = '';
      return;
    };
  };

  const handleUsernameSubmit = (e: any): void => {
    e.preventDefault();

    const input = (document.querySelector('.input-add-username') as HTMLInputElement).value,
          target = (document.querySelector('.input-add-username') as HTMLInputElement);

    if (filter.isProfane(input)) {
      alert('we don\'t allow profane language');
      return;
    } else if (!filter.isProfane(input) && target.validity.valid && input.match(usernameFormat)) {
      createUserInstanceAfterGoogleSignIn(input);
      return;
    } else {
      alert('Your username does not match our rules of: 1) 3-12 lowercase letters or 3-12 lowercase letters and 2-4 numbers, 2) no symbols, 3) no uppercase characters, 4) no swear words');
      return;
    };

  };

  const handleReturnToHome = (e: any): void => {
    e.preventDefault();
    toggleAddUsernamePage();
  };

  return (
    <form
      className="add-username-form"
      onSubmit={(e) => handleUsernameSubmit(e)}>
      <button
        className="return-from-username-button"
        type="button"
        onClick={(e) => handleReturnToHome(e)}>
        Return to Main
      </button>
      <label
        className="label-add-username">
        **Username:
      </label>
      <input
        className="input-add-username"
        placeholder="MasterChief91"
        minLength={3}
        maxLength={16}
        type="text"
        required
        onChange={() => validateUsername()}>
      </input>
      <button
        className="submit-username-button"
        type="submit" >
        Submit
      </button>
    </form>
  );
};

export default AddUsername;