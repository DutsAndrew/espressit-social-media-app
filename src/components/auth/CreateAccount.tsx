import React, { FC, MouseEventHandler, KeyboardEvent } from "react";
import { CreateAccountProps } from '../../types/interfaces';
const Filter = require('bad-words');

const CreateAccount: FC<CreateAccountProps> = (props): JSX.Element => {

  const { createAccountWithEmailAndPassword, handleSignUp } = props;

  const usernameFormat: RegExp = /^[a-z]{3,12}$|^[a-z]{3,12}\d{2,4}$/g;
  const mailFormat: RegExp = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/g;
  const passwordFormat: RegExp = /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/g;
  const filter = new Filter();

  const handleFormChange = (e: any): void => {
    const entryThatChanged = e.target as HTMLInputElement;
    const errorText = e.target.nextSibling;
    if (entryThatChanged && errorText) {
      if ((entryThatChanged.validity.valid && entryThatChanged.value.match(usernameFormat) && !filter.isProfane(entryThatChanged.value))
        || (entryThatChanged.validity.valid && entryThatChanged.value.match(mailFormat))
        || (entryThatChanged.validity.valid && entryThatChanged.value.match(passwordFormat))
      ) {
        errorText.textContent = "";
        errorText.className = "error";
      } else {
        showError(entryThatChanged, errorText);
      };
    };
  };

  const showError = (entry: any, error: any): void => {

    if (entry.id === 'username-input') {
      if (entry.validity.valueMissing) {
        error.textContent = "You must have a username to create an account";
        error.classList.add("error", "error-active");
      } else if (!entry.value.match(usernameFormat)) {
        error.textContent = "Your username does not match our rules of: 1) 3-12 lowercase letters or 3-12 lowercase letters and 2-4 numbers, 2) no symbols, 3) no uppercase characters";
        error.classList.add("error", "error-active");
      } else if (filter.isProfane(entry.value)) {
        error.textContent = "We don't allow profanity in usernames, sorry :(";
        error.classList.add("error", "error-active");
      };
      return;
    };

    if (entry.id === 'email-input') {
      if (entry.validity.valueMissing) {
        error.textContent = "You need to enter an email address";
        error.classList.add("error", "error-active");
      } else if (!entry.value.match(mailFormat)) {
        error.textContent = `Your email address doesn't seem to follow the traditional email patterns, please try again`;
        error.classList.add("error", "error-active");
      } else if (entry.validity.tooShort) {
        error.textContent = `Your email should be at least ${entry.minLength} characters; you entered: ${entry.value.length}`;
        error.classList.add("error", "error-active");
      } else if (entry.validity.tooLong) {
        error.textContent = `Your email should be no more than ${entry.maxLength} characters; you entered: ${entry.value.length}`;
        error.classList.add("error", "error-active");
      }
      return;
    };

    if (entry.id === 'password-input') {
      if (entry.validity.valueMissing) {
        error.textContent = "You need to enter a password";
        error.classList.add("error", "error-active");
      } else if (!entry.value.match(passwordFormat)) {
        error.textContent = `Your password must have at least: 1) One uppercase letter, 2) One lowercase letter, 3) One number, 4) One symbol, 5) And be at least 8 characters in length`;
        error.classList.add("error", "error-active");
      } else if (entry.validity.tooShort) {
        error.textContent = `Your password should be at least ${entry.minLength} characters; you entered: ${entry.value.length}`;
        error.classList.add("error", "error-active");
      } else if (entry.validity.tooLong) {
        error.textContent = `Your password should be no more than ${entry.maxLength} characters; you entered: ${entry.value.length}`;
        error.classList.add("error", "error-active");
      }
      return;
    };

    if (entry.id === 'password-confirm-input') {
      if (entry.validity.valueMissing) {
        error.textContent = "You need to confirm your password";
        error.classList.add("error", "error-active");
      } else if (entry.value !== entry.previousSibling.previousSibling.value) {
        error.textContent = `Your passwords do not match`;
        error.classList.add("error", "error-active");
      }
      return;
    };
  };

  const submitAccountCreationForm = (e: any): void  => {
    e.preventDefault();

    const usernameEntry = (document.getElementById("username-input") as HTMLInputElement);
    const emailEntry = (document.getElementById("email-input") as HTMLInputElement);
    const passwordEntry = (document.getElementById("password-input") as HTMLInputElement);
    const passwordConfirmEntry = (document.getElementById("password-confirm-input") as HTMLInputElement);
    const activeErrors = document.querySelectorAll('.error-active').length;

    if (usernameEntry) {
      if (!usernameEntry.validity.valid || !usernameEntry.value.match(usernameFormat)) {
        showError(usernameEntry, usernameEntry.nextSibling);
        return;
      };
    };

    if (emailEntry) {
      if (!emailEntry.validity.valid || !emailEntry.value.match(mailFormat)) {
        showError(emailEntry, emailEntry.nextSibling);
        return;
      };
    };

    if (passwordEntry) {
      if (!passwordEntry.validity.valid || !passwordEntry.value.match(passwordFormat)) {
        showError(passwordEntry, passwordEntry.nextSibling);
        return;
      };
    };

    if (passwordConfirmEntry) {
      if ((!passwordConfirmEntry.validity.valid || !passwordConfirmEntry.value.match(passwordFormat))
        && passwordEntry.value === passwordConfirmEntry.value
        ) {
          showError(passwordConfirmEntry, passwordConfirmEntry.nextSibling);
          return;
      };
    };

    if (emailEntry.validity.valid
      && passwordEntry.validity.valid
      && passwordConfirmEntry.validity.valid
      && activeErrors === 0
      ) {
        createAccountWithEmailAndPassword(usernameEntry.value, emailEntry.value, passwordEntry.value);
      };
  };

  return (
    <form id="user-creation-form"
      onKeyDown={(e) => e.key === "Enter" ? e.preventDefault() : null} >
      <button id="close-form-button"
        onClick={handleSignUp as unknown as MouseEventHandler<HTMLButtonElement>} >
        X Close Form
      </button>
      <fieldset id="create-account-fieldset" >
        <legend id="create-account-legend">
          Create an account:
        </legend>
        <label htmlFor="username-input" 
          id="username-label">
          *Username:
        </label>
        <input id="username-input" 
          placeholder="MischievousJack92"
          onChange={handleFormChange}
          minLength={3}
          maxLength={16}
          type="text"
          required>
        </input>
        <p id="username-input-error"
          className ="error-msg" >
        </p>
        <label htmlFor="email-input"
          id="email-label" >
          *Email:
        </label>
        <input id="email-input" 
          placeholder="JohnWick92@gmail.com"
          onChange={handleFormChange}
          minLength={9}
          maxLength={253}
          type="email"
          required>
        </input>
        <p id="email-input-error"
          className ="error-msg" >
        </p>
        <label htmlFor="password-input"
          id="password-label" >
          *Password:
        </label>
        <input id="password-input"
          placeholder="********"
          onChange={handleFormChange}
          minLength={8}
          maxLength={127}
          type="password"
          data-testid="password-input"
          required>
        </input>
        <p id="password-input-error"
          className ="error-msg" >
        </p>
        <label htmlFor="password-confirm-input"
          id="password-confirm-label" >
          *Confirm Password:
        </label>
        <input id="password-confirm-input"
          placeholder="********"
          onChange={(e) => handleFormChange(e)}
          minLength={8}
          maxLength={127}
          type="password"
          data-testid="confirm-password-input"
          required>
        </input>
        <p id="password-confirm-input-error"
          className ="error-msg" >
        </p>
        <button id="account-submit"
          type="button"
          onClick={(e) => submitAccountCreationForm(e)}>
            Submit
        </button>
      </fieldset>
    </form>
  );
};

export default CreateAccount;