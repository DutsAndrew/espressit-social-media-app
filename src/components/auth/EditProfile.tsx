import React, { FC } from "react";
import { EditProfileProps } from '../../types/interfaces';

const EditProfile: FC<EditProfileProps> = (props): JSX.Element => {

  const { currentUser, toggleEditProfilePage } = props;

  console.log(currentUser);

  const handleProfileEdit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    console.log('editing profile');
  };

  const handleReturnToMain = () => {
    toggleEditProfilePage();
  };

  return (
    <form className="edit-profile-form" onSubmit={handleProfileEdit}>
      <button type="button" className="return-to-main-page-button" onClick={handleReturnToMain} >
        Return To Home
      </button>
      <fieldset className="edit-profile-fieldset">
        <legend className="edit-profile-legend" >Profile Information:</legend>
        <label htmlFor="first-name" className="edit-profile-label">First Name:</label>
        <input id="first-name-input" name="first-name" className="edit-profile-input" ></input>
        <label htmlFor="last-name" className="edit-profile-label">Last Name:</label>
        <input id="last-name-input" name="last-name" className="edit-profile-input" ></input>
        <label htmlFor="user-name" className="edit-profile-label">Username:</label>
        <input id="user-name-input" name="user-name" className="edit-profile-input" ></input>
        <label htmlFor="email" className="edit-profile-label">Email:</label>
        <input id="email-input" name="email" className="edit-profile-input" ></input>
        <button type="submit" className="submit-profile-button">
          Submit Information
        </button>
      </fieldset>
      <fieldset className="edit-profile-fieldset">
        <legend className="edit-profile-legend" >Danger Zone</legend>
        <button type="button" className="delete-account-button">
          Delete Account
        </button>
        <button type="button" className="remove-data-button">
          Remove all account information and posts
        </button>
      </fieldset>
    </form>
  );
};

export default EditProfile;