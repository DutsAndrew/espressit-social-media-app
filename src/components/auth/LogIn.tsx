import React, { FC } from "react";
import { LogInProps } from '../../types/interfaces';
import google from '../../assets/google-logo.svg';

const LogIn: FC<LogInProps> = (props): JSX.Element => {

  const { signInUser, handleLogIn, signInWithGoogleAccount } = props;

  const handleSignIn = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const emailEntry: any = document.querySelector('#email-login');
    const passwordEntry: any = document.querySelector('#password-login');
    if (emailEntry && passwordEntry) {
      if (emailEntry.validity.valid && passwordEntry.validity.valid) {
        signInUser(emailEntry.value, passwordEntry.value);
      };
    };
  };

  const handleGoogleSignIn = (): void => {
    signInWithGoogleAccount();
  };

  return (
    <form id="sign-in-form" onSubmit={handleSignIn} >
      <button type="button"
        id="close-form-button"
        onClick={handleLogIn}>
        X Close Form
      </button>
      <fieldset id="sign-in-fieldset" >
        <legend>Sign In:</legend>
        <label htmlFor="email-login" >
          *Email:
        </label>
        <input id="email-login"
          placeholder="JohnWick92@gmail.com"
          minLength={9}
          maxLength={253}
          type="email"
          required>
        </input>
        <label htmlFor="password-login" >
          *Password:
        </label>
        <input id="password-login"
          placeholder="********"
          minLength={8}
          maxLength={127}
          type="password"
          required>
        </input>
        <button id="sign-in-button"
          type="submit" >
          Sign In
        </button>
      </fieldset>
      <button type="button"
        id="google-sign-in-button"
        onClick={handleGoogleSignIn} >
        Sign in with 
        <img src={google}
          alt="google logo">
        </img>
      </button>
    </form>
  );
};

export default LogIn;