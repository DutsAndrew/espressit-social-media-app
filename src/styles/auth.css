.sign-up-link, .sign-in-link {
  margin: 0;
  text-decoration: underline;
  color: rgb(194, 161, 134);
}

.sign-up-link:hover, .sign-in-link:hover {
  cursor: pointer;
  font-size: 0.95em;
  color: rgb(138, 109, 76);
}

#close-form-button {
  width: 15vw;
  height: 3vh;
  margin-bottom: 4%;
  color: #e4bc84;
  background-color: #37251b;
}

/* CREATE ACCOUNT FORM */

#user-creation-form {
  display: grid;
  grid-row: 1 / 2;
  grid-column: 1 / 2;
  background-color: #ad6d2f;
  color: #37251b;
  width: 25vw;
  height: 50vh;
  font-size: 1.25em;
  justify-items: center;
  align-items: center;
  justify-content: center;
  align-content: center;
  margin-bottom: 1%;
}

#create-account-fieldset {
  display: flex;
  flex-flow: column nowrap;
}

#account-submit {
  width: 20vw;
  height: 3vh;
  color: #e4bc84;
  background-color: #37251b;
  margin-bottom: 1%;
}

/* SIGN IN FORM */

#sign-in-form {
  display: grid;
  grid-row: 1 / 2;
  grid-column: 3 / 4;
  background-color: #ad6d2f;
  color: #37251b;
  width: 25vw;
  height: 50vh;
  font-size: 1.25em;
  justify-items: center;
  align-items: center;
  justify-content: center;
  align-content: center;
  margin-bottom: 1%;
}

#sign-in-fieldset {
  display: flex;
  flex-flow: column nowrap;
  height: auto;
}

#email-login, #password-login {
  margin-bottom: 30%;
}

#sign-in-button {
  width: 20vw;
  height: 3vh;
  color: #e4bc84;
  background-color: #37251b;
  margin-bottom: 1%;
}

/* EXTERNAL LOGIN STYLES */

#external-login {
  display: grid;
  grid-row: 3;
  grid-column: 1 / 4;
  justify-items: center;
  align-items: start;
  align-content: center;
  justify-content: center;
}

#google-sign-in-button {
  width: 15vw;
  height: 4vh;
  color: #e4bc84;
  background-color: #37251b;
  margin-top: 4%;
  margin-bottom: 1%;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-evenly;
  align-items: center;
}

/* FORM VALIDATION */

input:invalid {
  border: 2px dashed red;
  box-shadow: 0 0 5px 1px red;
}

input:valid {
  border: 2px solid #37251b;
}

input:invalid:required {
  background-image: linear-gradient(to right, #e4bc84, #d2c1b0);
  box-shadow: 0 0 5px 1px #8f755f;
}

input:focus:invalid {
  outline: none;
}

.error {
  width  : 100%;
  padding: 0;

  font-size: 80%;
  color: white;
  background-color: #900;
  border-radius: 0 0 5px 5px;

  box-sizing: border-box;
}

.error-active {
  padding: 0.3em;
}