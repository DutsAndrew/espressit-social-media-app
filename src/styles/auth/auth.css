/* used in both Create Account and Log In */

@media (max-width: 600px) {
  #close-form-button {
    width: 40vw;
    height: 5vh;
    font-size: 1em;
    color: #e4bc84;
    background-color: #37251b;
  }
}

@media (min-width: 601px) {
  #close-form-button {
    width: 20vw;
    height: 5vh;
    font-size: 1em;
    color: #e4bc84;
    background-color: #37251b;
  }
}

/* SIGNUP AND SIGN IN TEXT STYLING */

.sign-up-link, .sign-in-link {
  margin: 0;
  text-decoration: underline;
  color: rgb(194, 161, 134);
}

.sign-up-link:hover, .sign-in-link:hover {
  cursor: pointer;
  font-size: 0.95em;
  color: rgb(138, 109, 76);
}

/* FORM VALIDATION */

input:invalid {
  border: 2px dashed red;
  box-shadow: 0 0 5px 1px red;
}

input:valid {
  border: 2px solid #37251b;
}

input:invalid:required {
  background-image: linear-gradient(to right, #e4bc84, #d2c1b0);
  box-shadow: 0 0 5px 1px #8f755f;
}

input:focus:invalid {
  outline: none;
}

.error {
  width  : 100%;
  padding: 0;

  font-size: 80%;
  color: rgb(253, 255, 246);
  background-color: #900;
  border-radius: 0 0 5px 5px;

  box-sizing: border-box;
}

.error-active {
  padding: 0.3em;
}