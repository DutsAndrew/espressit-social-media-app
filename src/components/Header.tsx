import React, { FC } from "react";
import styled from 'styled-components'
import '../styles/auth.css';
import espresso from '../assets/espresso.svg';
import { HeaderProps } from "../types/interfaces";
import AccountDisplay from "./auth/AccountDisplay";

const HeaderWrapper = styled.div `
  width: 100vw;
  height: 10vh;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  background-color: rgb(77, 39, 14, 1);
  border-bottom: 1px solid #80411e;
`;

const LogoWrapper = styled.div `
  width: 30vw;
  height: 10vh;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  gap: 1vw;
`;

const Title = styled.p `
  width: 5vw;
  height: 10vh;
  font-size: 2em;
  margin: 0;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  color: rgb(138, 109, 76);
`;

const AccountWrapper = styled.div `
  width: 30vw;
  height: 10vh;
  font-size: 1em;
  display: flex;
  margin: 0;
  flex-flow: column nowrap;
  justify-content: space-evenly;
  align-items: center;
`;

const Header: FC<HeaderProps> = (props): JSX.Element => {

  const { handleSignUp, handleLogIn, currentUser, signOut, toggleEditProfilePage, toggleViewFavoritesPage } = props;

  // if user isn't logged in
  if (typeof currentUser === 'string') {
    return (
      <HeaderWrapper>
        <LogoWrapper>
          <img className="espresso-logo" src={espresso} alt="espresso" ></img>
          <Title>
            Espressit!
          </Title>
        </LogoWrapper>
        <AccountWrapper>
          <p className="sign-up-link" onClick={handleSignUp} >Sign Up</p>
          <p className="sign-in-link" onClick={handleLogIn} >Sign In</p>
        </AccountWrapper>
      </HeaderWrapper>
    );
  };

  // if logged in
  return (
    <HeaderWrapper>
      <LogoWrapper>
        <img className="espresso-logo" src={espresso} alt="espresso" ></img>
        <Title>
          Espressit!
        </Title>
      </LogoWrapper>
      <AccountWrapper>
        <AccountDisplay currentUser={currentUser}
          signOut={signOut}
          toggleEditProfilePage={toggleEditProfilePage} 
          toggleViewFavoritesPage={toggleViewFavoritesPage}
        />
      </AccountWrapper>
    </HeaderWrapper>
  );
};

export default Header;