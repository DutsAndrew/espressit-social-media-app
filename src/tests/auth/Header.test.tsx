import { render, screen } from '@testing-library/react';
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import Header from '../../components/Header';
import { User } from "firebase/auth";

describe('unit test for header', () => {

  // props passed into <Header/>
  const signUpMock = jest.fn();
  const logInMock = jest.fn();
  const signOutMock = jest.fn();
  const editProfileMock = jest.fn();
  const toggleViewContributionsPageMock = jest.fn();
  const userMock = "Bob Dillon" as unknown as User;

  test('renders header, title', () => {

    render(
      <Header handleSignUp={signUpMock}
        handleLogIn={logInMock}
        currentUser={userMock}
        signOut={signOutMock}
        toggleEditProfilePage={editProfileMock}
        toggleViewContributionsPage={toggleViewContributionsPageMock}
      />
    );

    const websiteTitle = screen.getByText(/Espressit!/i);
    expect(websiteTitle).toBeInTheDocument();
  });

  test('header matches snapshot', () => {
    const { container } = render(
      <Header handleSignUp={signUpMock}
        handleLogIn={logInMock}
        currentUser={userMock}
        signOut={signOutMock}
        toggleEditProfilePage={editProfileMock}
        toggleViewContributionsPage={toggleViewContributionsPageMock}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('signup and login are visible when not signed in', () => {

    render(
      <Header handleSignUp={signUpMock}
        handleLogIn={logInMock}
        currentUser={userMock}
        signOut={signOutMock}
        toggleEditProfilePage={editProfileMock}
        toggleViewContributionsPage={toggleViewContributionsPageMock}
      />
    );

    const signUpText = screen.getByText(/Sign Up/i);
    const signInText = screen.getByText(/Sign In/i);
    expect(signUpText).toBeInTheDocument();
    expect(signInText).toBeInTheDocument();
  });

  test('sign up form opens on text click', () => {

    render(
      <Header handleSignUp={signUpMock}
        handleLogIn={logInMock}
        currentUser={userMock}
        signOut={signOutMock}
        toggleEditProfilePage={editProfileMock}
        toggleViewContributionsPage={toggleViewContributionsPageMock}
      />
    );

    const link: any = screen.getByText("Sign Up");
    userEvent.click(link);
    expect(signUpMock).toHaveBeenCalled();
  });

  test('sign in form opens on text click', () => {
    
    render(
      <Header handleSignUp={signUpMock}
        handleLogIn={logInMock}
        currentUser={userMock}
        signOut={signOutMock}
        toggleEditProfilePage={editProfileMock}
        toggleViewContributionsPage={toggleViewContributionsPageMock}
      />
    );

    const link: any = screen.getByText("Sign In");
    userEvent.click(link);
    expect(logInMock).toHaveBeenCalled();
  });

});