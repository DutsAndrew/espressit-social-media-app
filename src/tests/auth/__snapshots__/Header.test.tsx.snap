// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for header header matches snapshot 1`] = `
<div>
  <div
    class="sc-bcXHqe eGdZpx"
  >
    <div
      class="sc-gswNZR bhqVXH"
    >
      <img
        alt="espresso"
        class="espresso-logo"
        src="espresso.svg"
      />
      <p
        class="sc-dkrFOg cIRlKe"
      >
        Espressit!
      </p>
    </div>
    <div
      class="sc-hLBbgP daKoNG"
    >
      <p
        class="sign-up-link"
      >
        Sign Up
      </p>
      <p
        class="sign-in-link"
      >
        Sign In
      </p>
    </div>
  </div>
</div>
`;
